File: py_src/problem003.py
# Project Euler Problem 2 Solution
#
# Problem statement:
# The prime factors of 13195 are 5, 7, 13 and 29. What
# is the largest prime factor of the number 600851475143?
#
# Solution description:
# Brute-force.
#
# Author: Daniel Schuette
# Date: 2018/09/30
# License: MIT (see ../LICENSE.md)
import math
import time

import numpy as np


def is_prime(n):
    """
    This function returns a boolean value indicating whether
    the input `n' is a primer number or not.
    """
    # handle edge cases
    if n == 1:
        return False
    elif n == 2:
        return True
    elif n % 2 == 0:
        return False

    for i in range(int(math.sqrt(n)+1)):
        if (i == 0) or (i == 1):
            continue
        if (n % i) == 0:
            return False
    return True


def largest_prime_factor(n, div=2):
    """
    This function returns the largest prime factor of
    input integer `n'.
    """
    target = n
    i = n // div  # the largest prime factor must be smaller than n//div
    while True:
        i -= 1
        if i <= 1:
            break  # pragma: no cover
        if (target % i) == 0:
            if is_prime(i):
                return i


if __name__ == "__main__":
    start = time.time()
    target = 600851475143
    solution = largest_prime_factor(target, div=2900)
    end = time.time()

    # `largest_prime_factor' implements the brute-force solution
    print("expired: {}s, solution: {}".format(
        np.round(end - start), solution))

File: rust_src/src/lib/problem003.rs
/* Project Euler Problem 3 Solution
 *
 * Problem statement:
 * The prime factors of 13195 are 5, 7, 13 and 29. What
 * is the largest prime factor of the number 600851475143?
 *
 * Solution description:
 * --
 *
 * Author: Daniel Schuette
 * Date: 2019/02/27
 * License: MIT (see ../LICENSE.md)
 */
use crate::utils;

pub fn largest_prime_factor(n: i64) -> i64 {
    let mut divisor = 2;
    let mut current;

    loop {
        // calculate number to test
        if n % divisor == 0 {
            current = n / divisor;
        } else {
            divisor += 1; /* increment divisor once more */
            if divisor > n {
                return -1; /* no prime factor found */
            }
            continue; /* continue to next iteration */
        }

        // test primality and potentially return
        if utils::is_prime(current) {
            return current;
        }
        divisor += 1; /* increment for next iteration */
    }
}
